src/sick_mrs1000_parser.cpp

```diff
@@ -0,0 +1,346 @@
+
+  // ----- read fields into scan
+  ROS_DEBUG_STREAM("publishing with frame_id " << scan.header.frame_id);
+
+  ros::Time start_time = ros::Time::now(); // will be adjusted in the end
+
+  // 16: Scanning Frequency (5DC)
+  unsigned short scanning_freq = -1;
+  sscanf(fields[16], "%hx", &scanning_freq);
+  scan.scan_time = 1.0 / (scanning_freq / 100.0);
+  // ROS_DEBUG("hex: %s, scanning_freq: %d, scan_time: %f", fields[16], scanning_freq, scan.scan_time);
+
+  // 17: Measurement Frequency (36)
+  unsigned short measurement_freq = -1;
+  sscanf(fields[17], "%hx", &measurement_freq);
+  scan.time_increment = 1.0 / (measurement_freq * 100.0);
+  if (override_time_increment_ > 0.0)
+  {
+    // Some lasers may report incorrect measurement frequency
+    scan.time_increment = override_time_increment_;
+  }
+
+  // 23: Starting angle (FFF92230)
+  int starting_angle = -1;
+  sscanf(fields[23], "%x", &starting_angle);
+  scan.angle_min = (starting_angle / 10000.0) / 180.0 * M_PI - M_PI / 2;
+  // ROS_DEBUG("starting_angle: %d, angle_min: %f", starting_angle, scan.angle_min);
+
+  // 24: Angular step width (2710)
+  unsigned short angular_step_width = -1;
+  sscanf(fields[24], "%hx", &angular_step_width);
+  scan.angle_increment = (angular_step_width / 10000.0) / 180.0 * M_PI;
+  scan.angle_max = scan.angle_min + (number_of_data - 1) * scan.angle_increment;
+
+  // 25: Number of data (<= 10F)
+  // This is already determined above in number_of_data
+
+  // adjust angle_min to min_ang config param
+  int index_min = 0;
+  while (scan.angle_min + scan.angle_increment < config.min_ang)
+  {
+    scan.angle_min += scan.angle_increment;
+    index_min++;
+  }
+
+  // adjust angle_max to max_ang config param
+  int index_max = number_of_data - 1;
+  while (scan.angle_max - scan.angle_increment > config.max_ang)
+  {
+    scan.angle_max -= scan.angle_increment;
+    index_max--;
+  }
+
+  ROS_DEBUG("index_min: %d, index_max: %d", index_min, index_max);
+  // ROS_DEBUG("angular_step_width: %d, angle_increment: %f, angle_max: %f", angular_step_width, scan.angle_increment, scan.angle_max);
+
+  double phi = scan.angle_min;
+  double alpha = (-layer/ 100) * M_PI / 180;
+
+  // first layer
+  // order of layers: 0, -250, 250, -500
+  if(layer_count_ == 0){
+    modifier_.resize(4 * (index_max - index_min + 1));
+    point_counter_ = 0;
+    x_iter = sensor_msgs::PointCloud2Iterator<float>(cloud_, "x");
+    y_iter = sensor_msgs::PointCloud2Iterator<float>(cloud_, "y");
+    z_iter = sensor_msgs::PointCloud2Iterator<float>(cloud_, "z");
+  }
+  layer_count_++;
+
+  // 26..26 + n - 1: Data_1 .. Data_n
+  scan.ranges.resize(index_max - index_min + 1);
+  for (int j = index_min; j <= index_max; ++j)
+  {
+    if(point_counter_ >= modifier_.size()){
+      ROS_WARN_STREAM("Resized point buffer from " << modifier_.size() << " to " << modifier_.size() + 100);
+      modifier_.resize(modifier_.size() * 100);
```

Comment:
Ugh, that check is ugly. Just for clarification, this is why it's necessary (you probably know, but I'm writing this down partly so I don't forget): Since the 4 layers are accumulated, the reconfigure callback can change the value of config.min_ang and max_ang and thereby index_min and index_max. This is why we can end up overrunning the buffer that we initially reserved for the point cloud, so we have to check here. All of these problems would be gone if we publish separate point clouds for each layer (see first comment for this file).